klippy/extras/firmware_retraction.py

```diff
@@ -42,33 +66,53 @@ def cmd_SET_RETRACTION(self, gcmd):
                                               self.unretract_speed, minval=1)
         self.unretract_length = (self.retract_length
                                  + self.unretract_extra_length)
+        self.lift_z = gcmd.get_float('LIFT_Z', self.lift_z, minval=0.)
         self.is_retracted = False
+        self.current_lift_z = 0.
+        self._override_gcode_move(undo=True)
     cmd_GET_RETRACTION_help = ("Report firmware retraction paramters")
     def cmd_GET_RETRACTION(self, gcmd):
         gcmd.respond_info("RETRACT_LENGTH=%.5f RETRACT_SPEED=%.5f"
                           " UNRETRACT_EXTRA_LENGTH=%.5f UNRETRACT_SPEED=%.5f"
+                          " LIFT_Z=%.5f"
                           % (self.retract_length, self.retract_speed,
-                             self.unretract_extra_length, self.unretract_speed))
+                             self.unretract_extra_length, self.unretract_speed,
+                             self.lift_z))
+
+    def cmd_G0(self, gcmd):
+        self.prev_G0(gcmd)
+        self._check_z_lift(gcmd)
+    def cmd_G1(self, gcmd):
+        self.prev_G1(gcmd)
+        self._check_z_lift(gcmd)
```

Comment:
I'm explicitly overwriting G0 and G1 separately here even though their basic klipper implementation is identical since users might redefine them separately using macros.